ouroboros-consensus-diffusion/src/ouroboros-consensus-diffusion/Ouroboros/Consensus/Node/GSM.hs

```diff
@@ -0,0 +1,330 @@
+
+    blockUntilCaughtUp :: m ()
+    blockUntilCaughtUp = do
+        varsCandidate <- atomically $ do
+            -- STAGE 1: all ChainSync clients report no subsequent headers
+            idlers        <- getChainSyncIdlers
+            varsCandidate <- getChainSyncCandidates
+            check $
+                               0  < Map.size    varsCandidate
+              && Set.size idlers == Map.size    varsCandidate
+              &&          idlers == Map.keysSet varsCandidate
+
+            -- STAGE 2: no candidate is better than the node's current
+            -- selection
+            candidates <- traverse StrictSTM.readTVar varsCandidate
+               -- TODO Simon Marlow wrote "Never read an unbounded number of
+               -- TVars in a single transaction because the O(n) performance of
+               -- readTVar then gives O(n*n) for the whole transaction." Is it
+               -- possible that varsCandidate will be big enough that this STM
+               -- transaction won't terminate even when the node is actually
+               -- done syncing?
+            selection  <- getCurrentSelection
+            let ok candidate =
+                    WhetherCandidateIsBetter False
+                 == candidateOverSelection selection candidate
+            check $ all ok candidates
+
+            pure varsCandidate
+
+        -- STAGE 3: the previous check wasn't so slow that the peer set changed
+        --
+        -- If the node is actually done syncing, then this should rarely fail.
+        --
+        -- TODO is this even unnecessary?
+        varsCandidate' <- atomically getChainSyncCandidates
+        when (Map.keys varsCandidate /= Map.keys varsCandidate') $ do
+            blockUntilCaughtUp
+
+{-------------------------------------------------------------------------------
+  A helper for constructing a real 'GsmView'
+-------------------------------------------------------------------------------}
+
+newtype WrapDurationUntilTooOld m blk = DurationUntilTooOld {
+    getDurationUntilTooOld ::
+         HardFork.Summary (HardForkIndices blk)
+      -> Slot.WithOrigin Slot.SlotNo
+      -> m DurationFromNow
+  }
+
+-- | The real system's 'durationUntilTooOld'
+realDurationUntilTooOld ::
+     Monad m
+  => NominalDiffTime
+     -- ^ If the volatile tip is older than this, then the node will exit the
+     -- @CaughtUp@ state.
+     --
+     -- Eg 'Ouroboros.Consensus.Node.llrnMaxCaughtUpAge'
+  -> Clock.SystemTime m
+  -> WrapDurationUntilTooOld m blk
+realDurationUntilTooOld maxCaughtUpAge systemTime =
+    DurationUntilTooOld $ \hfSummary woSlot -> do
+        now <- Clock.systemTimeCurrent systemTime
+        pure $ case woSlot of
+            Slot.Origin  -> toDur now $ Clock.RelativeTime 0
+            Slot.At slot ->
+                let qry = Qry.slotToWallclock slot
+                      -- TODO cache this in the KnownIntersectionState? Or even
+                      -- in the LedgerDB?
```

Comment:
We now have this comment in several places (here, it should probably not refer to `KnownIntersectionState`, but the basic point about caching still stands), it might make sense to think about factoring this out at some point.

Reply:
In this instance, I was able to use `runWithCachedSummary`.